File: python/src/jan/week1/boj/포항항_박재환.py
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
arr = [['X'] * (m + 2) for _ in range(n+2)]

for i in range(1, n+1):
    row = list(input().strip())
    for j in range(m):
        arr[i][j+1] = row[j]

stores = []
def map_arr():
    global arr

    store_idx = 1;
    for i in range(1, n+1):
        for j in range(1, m+1):
            if arr[i][j] == 'S':
                stores.append([i, j])
                arr[i][j] = 0
            elif arr[i][j] == 'K':
                stores.append([i, j])
                arr[i][j] = store_idx
                store_idx+=1

def get_time_table():
    time_table = [[-1_000_000]*len(stores) for _ in range(len(stores))]
    for store_idx in range(len(stores)):
        find_min_time(store_idx, time_table)
    return time_table

dx = [0,1,0,-1]
dy = [1,0,-1,0]
def find_min_time(store_idx, time_table):
    from collections import deque
    q = deque()

    visited = [[False]*(m+2) for _ in range(n+2)]
    x, y = stores[store_idx]
    q.append([x, y, 0])
    visited[x][y] = True

    while q:
        cur = q.popleft()
        cur_x = cur[0]
        cur_y = cur[1]
        cur_time = cur[2]

        if arr[cur_x][cur_y] != '.' and arr[cur_x][cur_y] != 'X':
            time_table[store_idx][arr[cur_x][cur_y]] = cur_time

        for d in range(4):
            nx = cur_x + dx[d]
            ny = cur_y + dy[d]

            if arr[nx][ny] == 'X' or visited[nx][ny]:
                continue
            visited[nx][ny] = True
            q.append([nx, ny, cur_time + 1])

min_time = 1_000_001
def solution():
    map_arr()
    time_table = get_time_table()
    get_permutation(0, [0]*len(stores), [False]*len(stores), 0, time_table)
    print(min_time)

def get_permutation(selected, seq, used, time, time_table):
    global min_time
    if selected == 6:
        min_time = min(min_time, time)
        return

    for i in range(0, len(stores)):
        if used[i]:
            continue

        time += time_table[seq[selected-1]][i]
        seq[selected] = i
        used[i] = True
        if time < min_time:
            get_permutation(selected+1, seq, used, time, time_table)
        time -= time_table[seq[selected - 1]][i]
        used[i] = False
if __name__ == '__main__':
    solution()

File: java/src/jan/week1/boj/포항항_박재환.java
package jan.week1.boj;

import java.util.*;
import java.io.*;

public class 포항항_박재환 {
    static BufferedReader br;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        init();
        br.close();
    }
    static StringTokenizer st;
    static int[] dx = {0,1,0,-1};
    static int[] dy = {1,0,-1,0};
    static int n, m;
    static char[][] map;
    static ArrayDeque<int[]> stores;      // s : 0, 이외 1 based
    static int minTime, storeCount;
    static void init() throws IOException {
        minTime = Integer.MAX_VALUE;
        stores = new ArrayDeque<>();
        st = new StringTokenizer(br.readLine().trim());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        map = new char[n+2][m+2];
        for(int x=0; x<n+2; x++) Arrays.fill(map[x], 'X');

        char storeIdx = '1';
        for(int x=1; x<n+1; x++) {
            String input = br.readLine().trim();
            for(int y=1; y<m+1; y++) {
                map[x][y] = input.charAt(y-1);
                if(map[x][y] == 'S') {
                    stores.offerFirst(new int[] {x, y});
                    map[x][y] = '0';
                }
                else if(map[x][y] == 'K') {
                    stores.offerLast(new int[] {x, y});
                    map[x][y] = storeIdx++;
                }
            }
        }
        storeCount = stores.size();
        findAllDist();
        int[] seq = new int[6];
        visitSeq(1, seq, new boolean[storeCount], 0);
        System.out.println(minTime == Integer.MAX_VALUE ? -1 : minTime);
    }

    static void visitSeq(int selected, int[] seq, boolean[] isUsed, int time) {
        if(selected == 6) {
            minTime = Math.min(minTime, time);
            return;
        }

        for(int i=1; i< storeCount; i++) {
            if(isUsed[i]) continue;
            if(allDist[seq[selected-1]][i] == -1) continue;
            time += allDist[seq[selected-1]][i];
            isUsed[i] = true;
            seq[selected] = i;
            if(time < minTime ) {
                visitSeq(selected+1, seq, isUsed, time);
            }
            isUsed[i] = false;
            time -= allDist[seq[selected-1]][i];
        }
    }

    static int[][] allDist;
    static void findAllDist() {
        allDist = new int[stores.size()][stores.size()];
        for(int i=0; i<stores.size(); i++) Arrays.fill(allDist[i], -1);
        for(int i=0; i<storeCount; i++) {    // i 번째 위치에서 모든 위치로의 거리
            allDist[i][i] = 0;
            findMinDist(i, stores.poll());
        }
    }

    static void findMinDist(int sIdx, int[] s) {
        Queue<int[]> q = new ArrayDeque<>();
        boolean[][] visited = new boolean[n+2][m+2];
        q.offer(new int[] {s[0], s[1], 0});
        visited[s[0]][s[1]] = true;

        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int curX = cur[0];
            int curY = cur[1];
            int curTime = cur[2];

            if(map[curX][curY] != 'X' && map[curX][curY] != '.') {
//                System.out.printf("sIdx : %d, toIdx : %d, curTime : %d\n", sIdx, map[curX][curY]-'0', curTime);
                allDist[sIdx][map[curX][curY]-'0'] = curTime;
            }

            for(int d=0; d<4; d++) {
                int nx = curX + dx[d];
                int ny = curY + dy[d];

                if(map[nx][ny] == 'X' || visited[nx][ny]) continue;

                q.offer(new int[] {nx, ny, curTime+1});
                visited[nx][ny] = true;
            }
        }
    }
}
